### ISSUE TITLE
Replacing lib/pq Copy In with Gorm

### ISSUE THREAD

**Issue by shannontan-bolt** [jackc/pgx#728]
Our stack utilizes Gorm v1.9.11, and we saw that lib/pq just updated to recommend using this library instead. We were able to successfully use pgx as a driver replacement for pq for most of Gorm's functionality, but one of the pieces that is missing is support for "Copy".

1. Gorm only exposes the ability to retrieve sql.Db when not in a transaction: https://github.com/jinzhu/gorm/blob/master/main.go#L134. Instead, you can retrieve a version of it that is scoped to the basic non-Tx functions: https://github.com/jinzhu/gorm/blob/master/interface.go#L9

2. It looks like pq's Copy functionality is incorporated into the Prepare call. https://github.com/lib/pq/blob/9eb3fc897d6fd97dd4aad3d0404b54e2f7cc56be/conn.go#L790
https://github.com/lib/pq/blob/9eb3fc897d6fd97dd4aad3d0404b54e2f7cc56be/copy.go#L64

And then, subsequent calls to Exec should go through the buffer.

In short: number 2 makes it such that copy functionality is available while using Gorm due to limitations in number 1.

I have a couple thoughts:
- 1) Expose ability to get PG conn for the purpose of copy in without creating a transaction (we're already in a transaction)
- 2) Emulate similar functionality as lib/pq to move it into prepare/exec.

For now, we won't be able to move to jackc/pgx since Gorm is too prevalent. I'll be filing a similar issue with Gorm and seeing if they are implementing any functionality to expose more of the underlying driver/connection (seems like they might not really be able to though).

**Comment by jackc** [comment]
I'm only a little familiar with lib/pq and even less so with Gorm so I probably don't totally understand what your describing , but I can at least explain how to use copy with pgx through `database/sql`.

`database/sql` doesn't directly expose a mechanism for DB specific functionality like copy. `lib/pq` worked around it with magic arguments to standard `database/sql` methods. `pgx` works around it by providing a means to go from a `*sql.DB` to a `*pgx.Conn` where the native copy functionality can be used.

Originally in pgx, you would use `stdlib.AcquireConn(db)`. This still works, but as of Go 1.13 `database/sql` provides `Conn.Raw()` which should enable the same function in a Go standard manner.

Not sure how any of that would be impacted by Gorm though.

**Comment by kynetiv** [comment]
@shannontan-bolt any chance you can share how you managed to replace pg with pgx ? Just looking into doing the same. 
> We were able to successfully use pgx as a driver replacement for pq

**Comment by shannontan-bolt** [comment]
@jackc - the issue here is that we maintain a Gorm instance instead of a connection object, so we were only able to access functionality to database/sql through interfaces that Gorm provided. These didn't quite line up, but since Gorm V2 uses pgx, I think I can close this out.

@kynetiv - tbh, replacing pgx with pq worked completely fine except for the copy in functionality that pq offered that Gorm exposed (albeit in a weird way).
